### DISCUSSION TITLE
Evaluating Replicad coming from OpenSCAD

### DISCUSSION THREAD

**Discussion by ch1z** [sgenoud/replicad#106] [Q&A] [answered]
Hi everyone!

I'm someone who uses OpenSCAD for 3D printing and considering switching to another CAD program for my purposes below:
1. Make my CAD programs available online and not require users to download a separate app
2. Protect my code as I have experienced people using my OpenSCAD code as if they're own and profiting from it despite the license I provide is for personal use only.

Looking at Replicad, it appears that it ticks all the boxes. However, though I was a programmer early in my career, I have not really worked so much on javascript and React js. I'm committed to learn Replicad but I want to make sure first if the library meets my needs.

My OpenSCAD programs normally deal with text/fonts and importing SVGs. You can take a look at my OpenSCAD programs in the link below to get a feel of what type of programs I use OpenSCAD for.

https://www.printables.com/search/models?o=latest&q=tag:openscad%20@chiz_m

Questions:
1. Do you think Replicad has enough features/commands comparable to OpenSCAD?
2. Does Replicad support generation of texts and extruding imported SVG/STL/3MF files?
3. Aside from React js, do I need to learn other languages or technologies? (I'm familiar with HTML, CSS, ASP)

TIA!

**Comment by raydeleu** [comment, marked as answer]
1) In my view Replicad has many features and compares well to Openscad. That said, there are things that Openscad is better in and things that Replicad is better in. Replicad is more CAD-like as it uses the Opencascade library as a basis. 

2) According to the API documentation it should be possible to import STL and STEP. However, the documentation notes that sometimes it will not work. I have not tested these functions. The import of SVG was discussed around a year ago. It was implemented in a version of the "bling my thing" app by the same programmer, but not yet in Replicad. However, in the discussions it is mentioned that he is planning to include it in the drawing library of Replicad. There is no import for 3MF files. 

As a workaround you could try to convert the SVG to STEP first with another converter such as 
https://github.com/iskwmsy/svg2step
and then try to import the STEP file. 

3) You would only need to learn the basics of javascript programming. I have learned this for CascadeStudio and have written a manual to use this tool. I have included a small introduction to javascript at the back of this manual https://github.com/raydeleu/CascadeStudioManual
.

**Reply by ch1z** [reply]
Thanks so much for taking the time to answer my questions! I will definitely look at the links you provided and hopes to learn from them!

I have started learning react js and hoping I come back to this thread with my first project running :)

**Reply by raydeleu** [reply]
I wonder why you would need to learn react.js to achieve the goals you mentioned. The workbench that is available at https://studio.replicad.xyz/workbench already has the feature that you can declare variables that can be changed by the user. I adapted the example in the documentation to add these variables on top of the code. When you now use the "settings" button on top of the window with the rendered shape, you will see a dialog where you can change these variables. You could do the same to enter the text or characters you want to add to one of your models. The only downside that I can think of is that the code is visible for the user, so your idea can be easily taken by others. But I think this is true as soon as you publish an example of your shape. Remodelling isn't that hard, so the real value is probably in the idea, not the code. 


<img width="1437" alt="image" src="https://github.com/sgenoud/replicad/assets/38007983/2e8ac718-bf29-4886-836a-a39ebde37187">

Have you also found my quick reference manual for Replicad at https://github.com/raydeleu/ReplicadManual/blob/main/replicad-qrc.adoc. I think that this could be a nice addition to the documentation that is available on the replicad website. Note that I wrote this some time ago, so in some cases the library has been changed. The most notable example is that the "Sketch" function (see Chapter 4 in my manual) has been changed to a "Draw" function (see Chapter 6 in my manual) although I think that "Sketch" still works.

Another source of information could be the examples given on the website but also in the "show and tell" part of the discussions.

**Reply by raydeleu** [reply]
I quickly performed a test with the code below, but it did not work ... "failed to load stl file". Using roughly the same code with a STEP file  resulted in the error message "This shape has not type, it is null"  

``` javascript

const main = ({importSTL},{}) => {

let step =  './Users/ray/Downloads/holder.stl'

  let testImport = importSTL(new Blob([step]))

  return testImport
};
```

I found the link to https://scriptcad.com/paulftw/hacklab-RPi-Case where it seems to work. The programmer of scriptcad.com has obviously included replicad as one of the kernels in his modelling website. However, building the model takes a very long time. 

<img width="1437" alt="image" src="https://github.com/sgenoud/replicad/assets/38007983/65f97a38-b1e4-4fd5-9d27-7884f0907a4c">

**Reply by ch1z** [reply]
I wanted to host my programs using my own domain and I want to hide my code. I want to have control on whoever can access my program. The workbench is a good tool/dev't environment as I test my programs but eventually I want the code hidden so I figured I need to learn react js so I can make that happen.

**Reply by ch1z** [reply]
Wow, you are right! It does take a while for the render to complete!

Btw, as I went into the rabbit hole, I also came across JSCAD. Link and sample below. Now, I'm not sure if it's the same or how it's different from Replicad. It appears that it's fast at rendering too.

https://openjscad.xyz/dokuwiki/doku.php?id=en:start 
https://bruceborrett.github.io/easy-enclosure/

Sorry, I know I'm going off a tangent here but I thought of sharing, wondering if you already know about this is not yet.

**Reply by raydeleu** [reply]
I have compared JSCAD with CascadeStudio in https://github.com/raydeleu/CascadeStudioManual#comparison-to-similar-tools. Replicad, just like CascadeStudio, uses a "real" CAD kernel that can be used to create almost any shape you can think of. The example of the splunge watering can that I created (see replicad documentation) illustrates this. JSCAD uses a javascript version of the OpenSCAD kernel. It creates objects based on polygons with limited resolution and operations like filleting are difficult to achieve. There is a script by Kurt Hutten (see https://www.youtube.com/watch?v=laxv2wFKq8Q) that comes close, but it just isn't the real thing, at least not for me. I approached the tools from an industrial design perspective and OpenSCAD/JSCAD do not come close to tools that are used for that purpose. If polygons were sufficient, I would rather use Blender.

**Reply by raydeleu** [reply]
If you want to build an application that hides your code, then Replicad is for you. Have a look at the example application by the author of replicad at https://blingmything.sgenoud.com/. 

<img width="1289" alt="image" src="https://github.com/sgenoud/replicad/assets/38007983/c41304f4-ce9f-4251-91b4-bf35f35a0665">

I think this is exactly what you intended. Judging from the code it is indeed built with react.

**Reply by sgenoud** [reply]
@raydeleu thanks for answering this! Life has been quite busy these last weeks (and the coming ones as well) - I will look more in detail into your question when I can find some time when I'm not knackered!)

**Reply by ch1z** [reply]
Thanks @raydeleu! I will definitely look at the example you provided!

**Comment by sgenoud** [comment]
@ch1z a quick note on working with STL with replicad. As it is now this is not the best - it will be very slow and cumbersome. 

There are two main different types of CAD models (and the software that goes with it). Mesh - which is what STL is and boundary representation (STEP files). As @raydeleu aluded to, replicad works best with boundary representation - while JSCAD (or Blender) works with meshes.

So depending on your use case you might want to go with one or the other!

**Reply by sgenoud** [reply]
> Aside from React js, do I need to learn other languages or technologies? (I'm familiar with HTML, CSS, ASP)

You don't need to learn that much about react - I use it to leverage the work made with react-three-fiber, which is a wrapper around three.js (you could work with raw three.js, or even raw WebGL or babylon.js).

**Reply by ch1z** [reply]
@sgenoud Thanks for the reply!

I just settled with STL coz that's the format that is most easily accessible by end users but since I will have more control over my code then perhaps I can adjust my code to accept STEP files instead.

As for learning react... It's still a long way for me to get up to speed but I'm getting there. I'll primarily use the knowledge to set up my future Replicad code hosted online on my website.
